confmap/provider/s3provider/provider_test.go

```diff
@@ -0,0 +1,201 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package s3provider
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/confmap"
+	"go.opentelemetry.io/collector/confmap/provider/internal"
+)
+
+// checkURI checks whether the s3-uri is valid
+func checkURI(uri string) error {
+	// check uri's prefix valid or not
+	if !strings.HasPrefix(uri, schemeName+":") {
+		return fmt.Errorf("%q uri is not supported by %q provider", uri, schemeName)
+	}
+	// Check if users set up their env for S3 Auth check yet
+	if os.Getenv("AWS_ACCESS_KEY_ID") == "" || os.Getenv("AWS_SECRET_ACCESS_KEY") == "" {
+		return fmt.Errorf("unable to fetch access keys for S3 Auth")
+	}
+	// check uri valid or not, should with 'Bucket, Region, Key'
+	_, _, _, err := S3URISplit(uri)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// testRetrieve: Mock how Retrieve() works in normal cases
+type testRetrieve struct{}
+
+func NewTestRetrieve() confmap.Provider {
+	return &testRetrieve{}
+}
+
+func (fp *testRetrieve) Retrieve(ctx context.Context, uri string, watcher confmap.WatcherFunc) (confmap.Retrieved, error) {
+	// check URI
+	err := checkURI(uri)
+	if err != nil {
+		return confmap.Retrieved{}, err
+	}
+	// read local config file and return
+	f, err := ioutil.ReadFile("../../testdata/config.yaml")
+	if err != nil {
+		return confmap.Retrieved{}, err
+	}
+	return internal.NewRetrievedFromYAML(f)
+}
+
+func (fp *testRetrieve) Scheme() string {
+	return schemeName
+}
+
+func (fp *testRetrieve) Shutdown(context.Context) error {
+	return nil
+}
+
+// testInvalidRetrieve: Mock how Retrieve() works when the returned config file is invalid
+type testInvalidRetrieve struct{}
+
+func NewTestInvalidRetrieve() confmap.Provider {
+	return &testInvalidRetrieve{}
+}
+
+func (fp *testInvalidRetrieve) Retrieve(ctx context.Context, uri string, watcher confmap.WatcherFunc) (confmap.Retrieved, error) {
```

Comment:
The mocking of the system under test here feels wrong.  If, instead, the `provider` struct includes a member that expects an interface including `GetObject(...)`, then a mock `s3.Client` implementation could be provided and the actual logic of the `provider.Retrieve()` method tested without needing to actually invoke S3.

Reply:
Thanks for telling me this bug, I just fixed this problem and pushed new codes!

At this time, I mock both that how S3 client and Retrieve() work.